### ISSUE TITLE
Null termination of particular string data types

### ISSUE THREAD

**Issue by AArnott** [microsoft/win32metadata#1513]
The docs are sadly inconsistent when it comes to defining a requirement for null termination.
I'm hoping folks can help me officially classify them in the table below. Then maybe we'll go file doc bugs.
It's important in the context of projections because we need to know how to marshal values that may not already be null-terminated.

Data type | Null terminated? | Notes
--|--|--
`PCWSTR`/`PWSTR` | [yes vote](https://learn.microsoft.com/en-us/windows/win32/winprog/windows-data-types) and [no vote](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/76f10dd8-699d-45e6-a53c-5aefc586da20).
`BSTR` | [yes](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-dtyp/692a42a9-06ce-4394-b9bc-5d2a50440168) | Also length prefixed, which allows for null to appear *within* the string as well. The .NET runtime for example allocates every string as a length prefixed and null-terminated string (for direct marshaling), though .NET strings are allowed to contain null characters.
`wchar_t*` | no?

**Comment by ChrisDenton** [comment]
I'm not sure there's a simple answer but if one is required then I'd say `PCWSTR` and `wchar_t*`, all must be null terminated. It's never wrong per se, or at least there's no harm in appending a null when one isn't required. The same should be assumed to be true for any similar type alias passed through the WIn32 API. At the end of the day, the only source of truth is the documentation for individual functions/methods (with a default assumption of "**must** be null terminated unless explicitly stated otherwise").

---

I guess the official answer would be from the headers:

```c
typedef wchar_t WCHAR;    // wc,   16-bit UNICODE character

typedef WCHAR *PWCHAR, *LPWCH, *PWCH;
typedef CONST WCHAR *LPCWCH, *PCWCH;

typedef _Null_terminated_ WCHAR *NWPSTR, *LPWSTR, *PWSTR;
typedef _Null_terminated_ PWSTR *PZPWSTR;
typedef _Null_terminated_ CONST PWSTR *PCZPWSTR;
typedef _Null_terminated_ WCHAR UNALIGNED *LPUWSTR, *PUWSTR;
typedef _Null_terminated_ CONST WCHAR *LPCWSTR, *PCWSTR;
typedef _Null_terminated_ PCWSTR *PZPCWSTR;
typedef _Null_terminated_ CONST PCWSTR *PCZPCWSTR;
typedef _Null_terminated_ CONST WCHAR UNALIGNED *LPCUWSTR, *PCUWSTR;

typedef _NullNull_terminated_ WCHAR *PZZWSTR;
typedef _NullNull_terminated_ CONST WCHAR *PCZZWSTR;
typedef _NullNull_terminated_ WCHAR UNALIGNED *PUZZWSTR;
typedef _NullNull_terminated_ CONST WCHAR UNALIGNED *PCUZZWSTR;
```

---

Personally, I wouldn't necessarily trust the headers to always use typdefs to properly distinguish between "this must be null terminated" and "this may omit the null if a length is given". But appending a null is never wrong whereas forgetting to is almost always wrong within the Win32 APIs[^1] so, yeah, just null terminated it.

[^1]: NT APIs often, but not always, take a `UNICODE_STRING` which do not require null termination.

**Comment by ChrisDenton** [comment]
tl;dr:
"...which MAY be null-terminated" is perhaps technically correct but somewhat unhelpful in the general case. Whereas a "pointer to a null-terminated string" is much closer to the truth even if there happen to be some APIs which reuse this type for (potentially) non-null terminated strings.

**Comment by riverar** [comment]
There's no general answer possible here; it's API specific. Some APIs may expect null termination, some the opposite, some may want a constant of a specific length where an optional NUL could cause error, etc. etc.

**Comment by ChrisDenton** [comment]
A constant of a specific length would be represented as a static array in a struct rather than a string pointer, no?

I mean, given the goal "we need to know how to marshal values that may not already be null-terminated" then I think the only general answer that can be given is to require null termination. More context is needed to omit the null.

**Comment by mikebattista** [comment]
There are `[NotNullTerminated]` and `[NullNullTerminated]` attributes in the metadata today. I believe they're added based on the header file definition [here](https://github.com/microsoft/win32metadata/issues/1513#issuecomment-1489193965). Is that helpful?

**Comment by JaiganeshKumaran** [comment]
Stop using custom string types. Just write wchar_t const* going forward. For not null terminated types, pass a size parameter along with the pointer.
